app/vminsert/common/insert_ctx.go

```diff
@@ -77,6 +108,7 @@ func (ctx *InsertCtx) WriteDataPointExt(metricNameRaw []byte, labels []prompbmar
 }
 
 func (ctx *InsertCtx) addRow(metricNameRaw []byte, timestamp int64, value float64) error {
+	ingestionRateLimiter.Register(1)
```

Comment:
I think L124 would be sporadic when the `Register` is called.  Also `FlushBufs` may be called elsewhere resetting the length of metricNameBuf.  I think the most efficient approach would be to modify each of the  callers of `WriteDataPoint` and `WriteDataPointExt` with the exception of streaming aggregation because the throttling will have already been applied.

Reply:
>  I think the most efficient approach would be to modify each of the callers of WriteDataPoint and WriteDataPointExt with the exception of streaming aggregation because the throttling will have already been applied.

Agree. Sounds like the most sensible way.
@f41gh7 do you have objections?

Reply:
I think it's fine to have approximate limit and apply it at `FlushBuffs`.
First of all, this limit should be applied to ingested samples and it actually happen during `FlushBuffs` call. 

  Yes, it may consume additional CPU for data parsing. And I suggest to check the following comment statement L184:
```
// There is no need in limiting the number of concurrent calls to vmstorage.AddRows() here,
	// since the number of concurrent FlushBufs() calls should be already limited via writeconcurrencylimiter
	// used at every stream.Parse() call under lib/protoparser/*
	err := vmstorage.AddRows(ctx.mrs)
```
 
 I think it's not true for `streaming` protocols. Such as `influx`, `csv`, `graphite`.

Reply:
How would `ingestionRateLimiter.Register(len(batchSize))` behave if `len(batchSize) > maxIngestionRate`?

Reply:
Consider the following example:
  maxIngestionRate=1_000  samples/per second
  total_request_batch=100_000
 1. flushRows at time=0s
  current_batch=10_000
  sample_budget=0
  total_request_batch=100_000
2. flushRows at time=1s
  current_batch=10_000
  sample_budget=-10_000
  total_request_batch=90_000
4. flushRows at time=2s
  current_batch=10_000
  sample_budget=-9_000
  total_request_batch=90_000
....
n.  flushRows at time=10s
  current_batch=10_000
  sample_budget=0
  total_request_batch=90_000
n+1. flushRows at time=11s
  current_batch=10_000
  sample_budget=-10_000
  total_request_batch=80_000

 Overall ingestion rate is at `1_000` samples per second for `10s`.

Reply:
Thanks @f41gh7 the above example makes sense. Are saying we should add the `Register(len(ctx.mrs))` before L184?

Reply:
Yes, it should be `FlushBufs`, sorry for typo.

Reply:
Nice! So I think consensus is to use limiter inside of FlushBufs function. This would be the least intrusive and most performance solution.